tests/components/vizio/test_media_player.py

```diff
@@ -0,0 +1,289 @@
+        return_value=vizio_power_state,
+    ):
+        await async_setup_component(hass, DOMAIN, {DOMAIN: config})
+        await hass.async_block_till_done()
+
+        attr = hass.states.get(ENTITY_ID).attributes
+        assert attr["friendly_name"] == NAME
+        assert attr["device_class"] == ha_device_class
+
+        assert hass.states.get(ENTITY_ID).state == ha_power_state
+        if ha_power_state == STATE_ON:
+            assert attr["source_list"] == INPUT_LIST
+            assert attr["source"] == CURRENT_INPUT
+            assert (
+                attr["volume_level"]
+                == float(int(MAX_VOLUME[vizio_device_class] / 2))
+                / MAX_VOLUME[vizio_device_class]
+            )
+
+
+async def test_vizio_speaker_on(hass: HomeAssistantType) -> None:
+    """Test for Vizio Speaker entity when on."""
+    await _test_vizio_init(
+        hass,
+        MOCK_SPEAKER_CONFIG,
+        VIZIO_DEVICE_CLASS_SPEAKER,
+        DEVICE_CLASS_SPEAKER,
+        True,
+        STATE_ON,
+    )
+
+
+async def test_vizio_speaker_off(hass: HomeAssistantType) -> None:
+    """Test for Vizio Speaker entity when off."""
+    await _test_vizio_init(
+        hass,
+        MOCK_SPEAKER_CONFIG,
+        VIZIO_DEVICE_CLASS_SPEAKER,
+        DEVICE_CLASS_SPEAKER,
+        False,
+        STATE_OFF,
+    )
+
+
+async def test_vizio_speaker_unavailable(hass: HomeAssistantType) -> None:
+    """Test for Vizio Speaker entity when unavailable."""
+    await _test_vizio_init(
+        hass,
+        MOCK_SPEAKER_CONFIG,
+        VIZIO_DEVICE_CLASS_SPEAKER,
+        DEVICE_CLASS_SPEAKER,
+        None,
+        STATE_UNAVAILABLE,
+    )
+
+
+async def test_vizio_init_tv_on(hass: HomeAssistantType) -> None:
+    """Test for Vizio TV entity when on."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        True,
+        STATE_ON,
+    )
+
+
+async def test_vizio_init_tv_off(hass: HomeAssistantType) -> None:
+    """Test for Vizio TV entity when off."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        False,
+        STATE_OFF,
+    )
+
+
+async def test_vizio_init_tv_unavailable(hass: HomeAssistantType) -> None:
+    """Test for Vizio TV entity when unavailable."""
+    await _test_vizio_init(
+        hass,
+        MOCK_USER_VALID_TV_CONFIG,
+        VIZIO_DEVICE_CLASS_TV,
+        DEVICE_CLASS_TV,
+        None,
+        STATE_UNAVAILABLE,
+    )
+
+
+async def test_setup_failure(hass: HomeAssistantType) -> None:
+    """Test media player entity setup failure."""
+    with patch(
+        "homeassistant.components.vizio.media_player.VizioAsync.can_connect",
+        return_value=False,
+    ):
+        try:
+            await async_setup_entry(hass, MOCK_SPEAKER_CONFIG_ENTRY, None)
```

Comment:
Use the core interfaces, either `async_setup_component` or `hass.config_entries.async_setup`. With the latter we need to create a mock entry first.

We can assert that no entity was added and use the `caplog` fixture if we want to assert the log warning.

Reply:
I don't need to assert the log warning, just that the entity wasn't added. But when I use `async_setup_component`, my patch does not result in the code block that fails the setup to be called. I will commit my latest code so you can see what I am trying to do

Reply:
I suspect it is because this check is failing during the config flow steps so that code block never gets reached. This is why I had called `homeassistant.components.vizio.media_player.async_setup_entry` directly, because `can_connect` needs to return `True` during config flow and `False` during entry setup to get to this block

Reply:
It's ok to set up the entry as I wrote above, via core interface. Create a mock entry first.

Reply:
I was able to resolve the issue with using the core `async_setup_component` by adding a fixture that I was using for Config Flow to ensure that the entry gets created. Assuming this is sufficient, this is resolved

Reply:
Disregard above, I moved to using entry setup